packages/server/src/shared/internal/serialize.ts

```diff
@@ -40,48 +40,83 @@ export type Serialize<T> =
 
 /** JSON serialize [tuples](https://www.typescriptlang.org/docs/handbook/2/objects.html#tuple-types) */
 type SerializeTuple<T extends [unknown, ...unknown[]]> = {
-  [k in keyof T]: T[k] extends NonJsonPrimitive ? null : Serialize<T[k]>;
+  [K in keyof T]: T[K] extends NonJsonPrimitive ? null : Serialize<T[K]>;
 };
 
 // prettier-ignore
-type SerializeObjectKey<T extends Record<any, any>, TKey> = 
+type SerializeObjectKey<T extends Record<any, any>, K> = 
   // never include entries where the key is a symbol
-  TKey extends symbol ? never : 
+  K extends symbol ? never : 
   // always include entries where the value is any
-  IsAny<T[TKey]> extends true ? TKey :
+  IsAny<T[K]> extends true ? K :
   // always include entries where the value is unknown
-  unknown extends T[TKey] ? TKey : 
+  unknown extends T[K] ? K : 
   // never include entries where the value is a non-JSON primitive
-  T[TKey] extends NonJsonPrimitive ? never : 
+  T[K] extends NonJsonPrimitive ? never : 
   // otherwise serialize the value
-  TKey;
+  K;
 /**
  * JSON serialize objects (not including arrays) and classes
  * @internal
  **/
 export type SerializeObject<T extends object> = {
-  [$Key in keyof T as SerializeObjectKey<T, $Key>]: Serialize<T[$Key]>;
+  [K in keyof T as SerializeObjectKey<T, K>]: Serialize<T[K]>;
 };
 
-type FilterDefinedKeys<TObj extends object> = Exclude<
+/**
+ * Extract keys from T where the value dosen't extend undefined
+ * Note: Can't parse IndexSignature or Record types
+ */
+type FilterDefinedKeys<T extends object> = Exclude<
   {
-    [TKey in keyof TObj]: undefined extends TObj[TKey] ? never : TKey;
-  }[keyof TObj],
+    [K in keyof T]: undefined extends T[K] ? never : K;
+  }[keyof T],
   undefined
 >;
 
-/*
- * For an object T, if it has any properties that are a union with `undefined`,
- * make those into optional properties instead.
- *
- * Example: { a: string | undefined} --> { a?: string}
+/**
+ * Get value of exactOptionalPropertyTypes config
+ */
+type ExactOptionalPropertyTypes = { a?: 0 | undefined } extends {
+  a?: 0;
+}
+  ? false
+  : true;
```

Comment:
That's a funky one. Can we just assume "strictest" and make sure we document it in tRPC instead of having these hacks?

Reply:
Hey, I agree about this being hacky, two options post removing it:
1. Allow incorrect types with `exactOptionalPropertyTypes=false`.
2. Generate merged type for both cases, looses type accuracy for literals when combined with `IndexSignature`.

In any case, I think a recommendation can be added in docs to use `exactOptionalPropertyTypes=true`. I tried turning it on in tRPC but the refactoring overhead is too much.

Let me know how to proceed and where to modify the docs. :)

Reply:
> In any case, I think a recommendation can be added in docs to use exactOptionalPropertyTypes=true. I tried turning it on in tRPC but the refactoring overhead is too much.

Ohh, we should have strictest in tRPC too - could you open an issue and one of the core team will do it?